# 01 — Pozos: perfilado del maestro de pozos (Capítulo IV)

**Objetivo:** conocer la estructura y la calidad del archivo **antes de limpiar nada**.
Este notebook solo *mide*: no modifica ni guarda datos.

| Dato | Valor |
| --- | --- |
| Archivo de origen | `data/raw/capitulo-iv-pozos.csv` |
| Fuente | Secretaría de Energía — ver `data/raw/README.md` |
| Qué contiene | Datos maestros de cada pozo: empresa, yacimiento, cuenca, provincia, profundidad, estado |
| Política aplicada | `POLITICA_CALIDAD_DATOS.md` — paso 1: **Perfilar** |
| Etapa siguiente | `02_pozos_limpieza.ipynb` — aplica las decisiones sobre estos hallazgos |

Las dimensiones de calidad que se revisan son: **completitud, unicidad, validez y consistencia**.

**Trazabilidad:** cada problema detectado recibe un código **H-xx** (hallazgo) en la tabla de la sección 10.
En `02_pozos_limpieza.ipynb`, cada regla de limpieza **R-xx** indica qué hallazgo resuelve.

## 1. Preparar el entorno

Importamos **pandas**, la librería de Python para trabajar con tablas.
Mostramos su versión porque forma parte de la reproducibilidad: con otra versión, algún resultado podría variar.

In [1]:
import pandas as pd

pd.set_option("display.max_columns", 50)
print("Versión de pandas:", pd.__version__)

Versión de pandas: 3.0.6


## 2. Cargar el archivo original

Leemos el CSV desde `data/raw`. Solo se **lee**: el archivo original no se toca.

- La ruta empieza con `../` porque el notebook está en la carpeta `notebooks/` y hay que "subir" un nivel.
- `low_memory=False` hace que pandas lea el archivo completo antes de decidir el tipo de cada columna, así evita tipos mezclados.

In [2]:
RUTA_RAW = "../data/raw/capitulo-iv-pozos.csv"

df = pd.read_csv(RUTA_RAW, low_memory=False)
print(f"Filas: {df.shape[0]:,}  |  Columnas: {df.shape[1]}")

Filas: 85,611  |  Columnas: 26


## 3. Estructura: columnas y tipos de datos

`info()` muestra cada columna, cuántos valores no vacíos tiene y su **tipo de dato**
(`int64` = número entero, `float64` = número con decimales, `str` = texto).

Es importante porque un tipo incorrecto trae problemas después: por ejemplo, las **fechas** vienen como texto
y así no se pueden ordenar ni restar.

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 85611 entries, 0 to 85610
Data columns (total 26 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   sigla                    85611 non-null  str    
 1   idpozo                   85611 non-null  int64  
 2   area                     85611 non-null  str    
 3   cod_area                 85611 non-null  str    
 4   empresa                  84647 non-null  str    
 5   yacimiento               85611 non-null  str    
 6   cod_yacimiento           85611 non-null  str    
 7   formacion                82796 non-null  str    
 8   cuenca                   85611 non-null  str    
 9   provincia                85611 non-null  str    
 10  cota                     85611 non-null  float64
 11  profundidad              85611 non-null  float64
 12  clasificacion            85611 non-null  str    
 13  subclasificacion         85611 non-null  str    
 14  tipo_recurso             85611 no

## 4. Primer vistazo a los datos

Miramos las primeras filas para entender qué representa cada columna.
Usamos `.T` (transpuesta) para ver las columnas como filas, porque son 26 y no entran a lo ancho.

In [4]:
df.head(3).T

,0,1,2
sigla,CH.CH.EaLE.x-1,CH.CH.EaLE.x-2,CH.CH.EaLE-3
idpozo,212,213,214
area,ESTANCIA LA ESCONDIDA,ESTANCIA LA ESCONDIDA,ESTANCIA LA ESCONDIDA
cod_area,ECH,ECH,ECH
empresa,COLHUE HUAPI S.A.,COLHUE HUAPI S.A.,COLHUE HUAPI S.A.
yacimiento,ESTANCIA LA ESCONDIDA,ESTANCIA LA ESCONDIDA,ESTANCIA LA ESCONDIDA
cod_yacimiento,ELA,ELA,ELA
formacion,comodoro rivadavia,comodoro rivadavia,comodoro rivadavia
cuenca,GOLFO SAN JORGE,GOLFO SAN JORGE,GOLFO SAN JORGE
provincia,Chubut,Chubut,Chubut


## 5. Completitud: ¿falta algún dato?

Contamos los **nulos** (celdas vacías) por columna y su porcentaje sobre el total.

Pero un dato puede faltar aunque la celda no esté vacía: esta fuente usa el texto **"No informado"**.
Son **nulos ocultos** y hay que contarlos aparte, porque `isna()` no los detecta.

In [5]:
completitud = pd.DataFrame({
    "nulos": df.isna().sum(),
    "no_informado": (df == "No informado").sum(),
})
completitud["faltantes_total"] = completitud["nulos"] + completitud["no_informado"]
completitud["pct_faltante"] = (completitud["faltantes_total"] / len(df) * 100).round(1)

completitud[completitud["faltantes_total"] > 0].sort_values("pct_faltante", ascending=False)

,nulos,no_informado,faltantes_total,pct_faltante
sub_tipo_recurso,0,80608,80608,94.2
adjiv_fecha_fin_term,36486,0,36486,42.6
adjiv_fecha_inicio_term,36487,0,36487,42.6
adjiv_fecha_fin_perf,34149,0,34149,39.9
adjiv_fecha_inicio_perf,34004,0,34004,39.7
tipo_recurso,0,21859,21859,25.5
subclasificacion,0,18199,18199,21.3
clasificacion,0,18199,18199,21.3
formacion,2815,0,2815,3.3
empresa,964,0,964,1.1


## 6. Unicidad: ¿hay registros repetidos?

Revisamos tres niveles:

1. **Filas completas duplicadas**: la misma fila exacta dos veces.
2. **`idpozo` repetido**: es la clave del maestro; si se repite, los cruces con los archivos de producción darían resultados duplicados.
3. **`sigla` repetida**: la sigla es el nombre del pozo. Si se repite con distinto `idpozo`, hay que entender por qué antes de decidir nada.

In [6]:
print("Filas completas duplicadas:", df.duplicated().sum())
print("idpozo repetidos:          ", df["idpozo"].duplicated().sum())

filas_sigla_repetida = df["sigla"].duplicated(keep=False)
print("Filas con sigla repetida:  ", filas_sigla_repetida.sum(),
      f"({filas_sigla_repetida.mean() * 100:.1f}% del total)")

Filas completas duplicadas: 0
idpozo repetidos:           0
Filas con sigla repetida:   13421 (15.7% del total)


Veamos algunos ejemplos de siglas repetidas para entender el caso.

In [7]:
df[filas_sigla_repetida].sort_values("sigla")[
    ["sigla", "idpozo", "empresa", "provincia", "tipoestado"]
].head(8)

,sigla,idpozo,empresa,provincia,tipoestado
83279,ACO.RN.TA-4001,164626,TANGO ENERGY ARGENTINA S.A.,Rio Negro,Extracción Efectiva
83280,ACO.RN.TA-4001,164627,TANGO ENERGY ARGENTINA S.A.,Rio Negro,Parado Transitoriamente
83326,ACO.RN.TA-4002,164673,TANGO ENERGY ARGENTINA S.A.,Rio Negro,Parado Transitoriamente
83327,ACO.RN.TA-4002,164674,TANGO ENERGY ARGENTINA S.A.,Rio Negro,Extracción Efectiva
83328,ACO.RN.TA-4002,164675,TANGO ENERGY ARGENTINA S.A.,Rio Negro,Parado Transitoriamente
63695,AEA.NQ.RCo.x-2001,133430,OILSTONE ENERGIA S.A.,Neuquén,Extracción Efectiva
81396,AEA.NQ.RCo.x-2001,162701,OILSTONE ENERGIA S.A.,Neuquén,Extracción Efectiva
76528,AEC.Nq.AC.a-2002,157635,OILSTONE ENERGIA S.A.,Neuquén,Extracción Efectiva


También revisamos si hay pozos distintos con **exactamente la misma ubicación** (columna `geojson`, que tiene las coordenadas).
Puede ser legítimo (por ejemplo, varios pozos desde una misma plataforma) o un error de carga.

In [8]:
filas_ubicacion_repetida = df["geojson"].duplicated(keep=False)
print("Filas que comparten ubicación con otro pozo:", filas_ubicacion_repetida.sum(),
      f"({filas_ubicacion_repetida.mean() * 100:.1f}%)")

Filas que comparten ubicación con otro pozo: 12104 (14.1%)


## 7. Validez: ¿los números tienen sentido?

`describe()` resume las columnas numéricas: mínimo, máximo, promedio y cuartiles.
Buscamos valores **imposibles o sospechosos**:

- **`profundidad`** (metros): un pozo no puede medir 0 m, y el más profundo del mundo ronda los 12.000 m.
- **`cota`** (metros sobre el nivel del mar): valores negativos o de 0 m en zonas de montaña son raros.

In [9]:
df[["cota", "profundidad"]].describe().round(1)

,cota,profundidad
count,85611.0,85611.0
mean,499.2,1717.8
std,308.5,1749.1
min,-100.0,0.0
25%,307.9,1032.0
50%,428.7,1595.0
75%,644.4,2352.0
max,5543.4,378939.0


Contamos los casos sospechosos y miramos los pozos más profundos.

In [10]:
print("Profundidad = 0:        ", (df["profundidad"] == 0).sum())
print("Profundidad > 10.000 m: ", (df["profundidad"] > 10_000).sum())
print("Cota negativa:          ", (df["cota"] < 0).sum())
print("Cota = 0:               ", (df["cota"] == 0).sum())

df.nlargest(7, "profundidad")[["idpozo", "sigla", "provincia", "profundidad"]]

Profundidad = 0:         10299
Profundidad > 10.000 m:  7
Cota negativa:           1
Cota = 0:                225


,idpozo,sigla,provincia,profundidad
75735,156804,YEA.RN.EFO-152(d),Rio Negro,378939.0
41544,109212,YPF.Ch.E.a-669,Chubut,27064.0
21214,69901,YPF.Md.PC-33,Mendoza,24223.0
34636,102326,YPF.SC.CG-441,Santa Cruz,20680.0
37495,105170,YPF.SC.CS-590,Santa Cruz,15273.0
4722,10269,TA.CMA-1.CullN-1,Tierra del Fuego,11021.0
29364,97069,YPF.SC.LP.a-2162,Santa Cruz,10600.0


### Validez de las fechas

Las cuatro columnas `adjiv_fecha_*` registran cuándo empezó y terminó la perforación y la terminación del pozo.
Vienen como **texto**, así que las convertimos a fecha **en una variable aparte**, sin tocar `df`, solo para medir:

- ¿Hay textos que no se pueden convertir a fecha?
- ¿Aparece la fecha **1900-01-01**? Suele usarse como "fecha de relleno" cuando no se conoce la real.
- ¿Hay fechas de **fin anteriores al inicio**?

In [11]:
columnas_fecha = [
    "adjiv_fecha_inicio_perf", "adjiv_fecha_fin_perf",
    "adjiv_fecha_inicio_term", "adjiv_fecha_fin_term",
]
fechas = df[columnas_fecha].apply(pd.to_datetime, errors="coerce")

resumen_fechas = pd.DataFrame({
    "vacias": df[columnas_fecha].isna().sum(),
    "no_convertibles": (fechas.isna() & df[columnas_fecha].notna()).sum(),
    "igual_1900_01_01": (fechas == "1900-01-01").sum(),
    "minima": fechas.min(),
    "maxima": fechas.max(),
})
resumen_fechas

,vacias,no_convertibles,igual_1900_01_01,minima,maxima
adjiv_fecha_inicio_perf,34004,0,37,1900-01-01,2026-05-06
adjiv_fecha_fin_perf,34149,0,33,1900-01-01,2026-05-22
adjiv_fecha_inicio_term,36487,0,33,1900-01-01,2026-06-01
adjiv_fecha_fin_term,36486,0,33,1900-01-01,2026-06-12


In [12]:
print("Fin de perforación antes del inicio:",
      (fechas["adjiv_fecha_fin_perf"] < fechas["adjiv_fecha_inicio_perf"]).sum())
print("Fin de terminación antes del inicio:",
      (fechas["adjiv_fecha_fin_term"] < fechas["adjiv_fecha_inicio_term"]).sum())

Fin de perforación antes del inicio: 3
Fin de terminación antes del inicio: 0


## 8. Consistencia: ¿los valores se escriben siempre igual?

Para las columnas de **categorías** (valores que se repiten de una lista) miramos cuántos valores distintos hay y cuáles son.
Así aparecen variantes de escritura, mayúsculas/minúsculas mezcladas o categorías raras.

In [13]:
columnas_categoria = [
    "provincia", "cuenca", "clasificacion", "tipo_recurso",
    "sub_tipo_recurso", "tipopozo", "tipoestado", "gasplus",
]
for col in columnas_categoria:
    print(f"--- {col} ({df[col].nunique()} valores distintos)")
    print(df[col].value_counts(dropna=False).to_string(), "\n")

--- provincia (13 valores distintos)
provincia
Santa Cruz          23707
Chubut              22606
Neuquén             19497
Mendoza              8897
Rio Negro            5773
La Pampa             2726
Tierra del Fuego     1242
Salta                 965
Estado Nacional        75
Formosa                65
Jujuy                  46
San Juan                6
Córdoba                 6 

--- cuenca (10 valores distintos)
cuenca
GOLFO SAN JORGE    44390
NEUQUINA           33155
CUYANA              3741
AUSTRAL             3236
NOROESTE            1065
NORESTE               16
LOS BOLSONES           3
CAÑADON ASFALTO        2
ÑIRIHUAU               2
GENERAL LEVALLE        1 

--- clasificacion (5 valores distintos)
clasificacion
EXPLOTACION       57412
No informado      18199
EXPLORACION        5950
SERVICIO           4017
ALMACENAMIENTO       33 

--- tipo_recurso (5 valores distintos)
tipo_recurso
CONVENCIONAL       58320
No informado       21859
NO CONVENCIONAL     5006
SIN RESERVORIO   

### Nombres y códigos

Cada `area` debería tener un único `cod_area`, y cada `yacimiento` un único `cod_yacimiento`.
Si la cantidad de nombres y de códigos no coincide, algún nombre tiene más de un código (o al revés).
Revisamos también si los textos mezclan mayúsculas y minúsculas.

In [14]:
print("Áreas distintas:      ", df["area"].nunique(), " | códigos de área:      ", df["cod_area"].nunique())
print("Yacimientos distintos:", df["yacimiento"].nunique(), " | códigos de yacimiento:", df["cod_yacimiento"].nunique())

areas_con_varios_codigos = df.groupby("area")["cod_area"].nunique()
print("Áreas con más de un código:", (areas_con_varios_codigos > 1).sum())

for col in ["empresa", "formacion", "area", "yacimiento"]:
    valores = df[col].dropna()
    print(f"{col:11} -> todo en mayúsculas: {valores.str.isupper().mean() * 100:5.1f}%"
          f" | todo en minúsculas: {valores.str.islower().mean() * 100:5.1f}%")

Áreas distintas:       457  | códigos de área:       464
Yacimientos distintos: 1184  | códigos de yacimiento: 1316
Áreas con más de un código: 7
empresa     -> todo en mayúsculas:  96.5% | todo en minúsculas:   0.0%
formacion   -> todo en mayúsculas:   0.0% | todo en minúsculas:  99.9%


area        -> todo en mayúsculas:  99.9% | todo en minúsculas:   0.0%


yacimiento  -> todo en mayúsculas:  99.9% | todo en minúsculas:   0.0%


## 9. Relaciones entre columnas

Esta sección se agregó el **28/09/2026**, al construir el diccionario de datos (`docs/diccionario_pozos.csv`).
Para escribir la definición de cada columna hubo que comprobar cómo se relacionan unas con otras,
y aparecieron cuatro hallazgos nuevos (**H-19 a H-22**).
El **30/09/2026**, al investigar la decisión D-3 (siglas repetidas), se agregó **H-23**, al investigar D-6 (cota), **H-24**. El **01/10/2026**, al investigar D-8 (textos), **H-25** y **H-26**.

Se registran acá, y no solo en el diccionario, por trazabilidad: todo lo que el diccionario marca como
*Verificado en el dato* tiene que tener su evidencia en código.

### Coordenadas: ¿los pozos caen dentro de la Argentina?

La columna `geojson` guarda la ubicación como texto, con este formato: `{"type":"Point","coordinates":[longitud, latitud]}`.
Usamos la librería **json** para leer ese texto y separar la **longitud** (este-oeste) de la **latitud** (norte-sur).

Después comprobamos si cada punto cae dentro de una "caja" aproximada que contiene a la Argentina continental
y su plataforma marítima: latitud entre -56 y -21,7 y longitud entre -74 y -53.
Un punto fuera de la caja indica un error de carga.

In [15]:
import json

coordenadas = df["geojson"].map(lambda texto: json.loads(texto)["coordinates"])
longitud = coordenadas.str[0]
latitud = coordenadas.str[1]

dentro_de_argentina = latitud.between(-56, -21.7) & longitud.between(-74, -53)
coordenadas_fuera = ~dentro_de_argentina
print("Pozos con coordenadas fuera de la Argentina:", coordenadas_fuera.sum())

df.loc[coordenadas_fuera, ["idpozo", "sigla", "provincia"]].assign(longitud=longitud, latitud=latitud)

Pozos con coordenadas fuera de la Argentina: 2


,idpozo,sigla,provincia,longitud,latitud
4598,10143,SJ.RN.LN-7,Rio Negro,-38.756338,-67.642383
80590,162058,PBE.Nq.M-1063(d),Neuquén,-38.590070,-69.415770


Los dos pozos tienen los valores **invertidos**: la longitud quedó en el lugar de la latitud y al revés.
Con los valores intercambiados, el punto cae en Río Negro y en Neuquén, que coincide con la columna `provincia`.
Con los valores como están, cae en la Antártida. → **H-19**

### Ubicación: ¿`geom` y `geojson` son la misma información?

La columna `geom` guarda la ubicación en **WKB**, un formato binario que usan las bases de datos geográficas, escrito en hexadecimal.
Lo decodificamos para compararlo con `geojson`:

- los caracteres de la posición 5 a 8 son el **SRID**, el código del sistema de coordenadas (4326 = WGS84, el del GPS);
- los siguientes 16 caracteres son la longitud y la latitud.

Si coinciden en todas las filas, `geom` es redundante y en la limpieza se puede evaluar conservar solo una de las dos.

In [16]:
import struct

def punto_desde_wkb(texto_hex):
    datos = bytes.fromhex(texto_hex)
    srid = struct.unpack("<I", datos[5:9])[0]
    x, y = struct.unpack("<dd", datos[9:25])
    return [srid, x, y]

puntos_geom = df["geom"].map(punto_desde_wkb)
iguales = ((puntos_geom.str[1] - longitud).abs() < 1e-9) & ((puntos_geom.str[2] - latitud).abs() < 1e-9)

print("Sistemas de coordenadas en geom:", puntos_geom.str[0].unique())
print("Filas donde geom coincide con geojson:", iguales.sum(), "de", len(df))

Sistemas de coordenadas en geom: [4326]
Filas donde geom coincide con geojson: 85611 de 85611


### Tipo de recurso y subtipo: ¿"No informado" es siempre un dato faltante?

`sub_tipo_recurso` dice si un pozo **no convencional** es **SHALE** (el hidrocarburo sigue en la roca madre, como en Vaca Muerta)
o **TIGHT** (arenas muy compactas). Para un pozo convencional, la pregunta no aplica.

Cruzamos las dos columnas con `pd.crosstab`, que cuenta cuántos pozos hay en cada combinación.

In [17]:
pd.crosstab(df["tipo_recurso"], df["sub_tipo_recurso"])

sub_tipo_recurso,No informado,SHALE,TIGHT
tipo_recurso,,,
CONVENCIONAL,58319,0,1
NO CONVENCIONAL,4,3380,1622
NO DISCRIMINADO,5,0,0
No informado,21859,0,0
SIN RESERVORIO,421,0,0


La tabla muestra que SHALE y TIGHT aparecen casi solo en los pozos NO CONVENCIONAL.
Entonces, "No informado" tiene **dos significados** según el tipo de recurso:

- en un pozo **CONVENCIONAL** significa **"no aplica"**: no es un dato faltante → **H-20**, que corrige la lectura de H-05;
- en un pozo **NO CONVENCIONAL** es un **faltante real**.

Además aparece un pozo CONVENCIONAL con subtipo informado, que contradice la jerarquía → **H-21**.

Los pozos con `tipo_recurso` SIN RESERVORIO, NO DISCRIMINADO o "No informado" quedan para la decisión de limpieza.

In [18]:
subtipo_no_aplica = (df["tipo_recurso"] == "CONVENCIONAL") & (df["sub_tipo_recurso"] == "No informado")
subtipo_faltante_real = (df["tipo_recurso"] == "NO CONVENCIONAL") & (df["sub_tipo_recurso"] == "No informado")
convencional_con_subtipo = (df["tipo_recurso"] == "CONVENCIONAL") & (df["sub_tipo_recurso"] != "No informado")

print("Convencionales con subtipo 'No informado' (no aplica):", subtipo_no_aplica.sum())
print("No convencionales sin subtipo (faltante real):        ", subtipo_faltante_real.sum())
print("Convencionales con subtipo informado (contradicción): ", convencional_con_subtipo.sum())

Convencionales con subtipo 'No informado' (no aplica): 58319
No convencionales sin subtipo (faltante real):         4
Convencionales con subtipo informado (contradicción):  1


### Yacimientos y áreas: ¿cada nombre tiene un solo código?

En la sección 8 vimos que hay 7 áreas con más de un código (H-17). Hacemos la misma comprobación para los yacimientos,
y también en el sentido contrario: si cada **código** tiene un único **nombre**.

In [19]:
codigos_por_yacimiento = df.groupby("yacimiento")["cod_yacimiento"].nunique()
yacimientos_con_varios_codigos = codigos_por_yacimiento[codigos_por_yacimiento > 1].index

print("Yacimientos con más de un código:        ", len(yacimientos_con_varios_codigos))
print("Pozos en esos yacimientos:               ", df["yacimiento"].isin(yacimientos_con_varios_codigos).sum())
print("Códigos de yacimiento con más de un nombre:", (df.groupby("cod_yacimiento")["yacimiento"].nunique() > 1).sum())
print("Códigos de área con más de un nombre:     ", (df.groupby("cod_area")["area"].nunique() > 1).sum())

Yacimientos con más de un código:         88
Pozos en esos yacimientos:                17478
Códigos de yacimiento con más de un nombre: 0
Códigos de área con más de un nombre:      0


Cada **código** corresponde a un solo nombre, pero hay nombres de yacimiento con varios códigos → **H-22**.
Una explicación posible es que el mismo nombre se use en áreas distintas; hay que investigarlo antes de decidir.

### Siglas repetidas: ¿son el mismo pozo con varias formaciones?

En la sección 6 vimos que 13.421 filas comparten `sigla` con otro `idpozo` (H-08).
Para entender por qué, agrupamos esas filas por sigla y medimos **en qué columnas cambia el valor** dentro de cada grupo.
Si lo único que cambia es la formación, se trata del mismo pozo físico declarado una vez por cada capa de roca que produce.

In [20]:
grupos_sigla = df[filas_sigla_repetida].groupby("sigla")
columnas_a_comparar = ["formacion", "geojson", "profundidad", "adjiv_fecha_inicio_perf", "tipoestado", "empresa"]
pct_distinto = pd.Series({
    col: (grupos_sigla[col].nunique(dropna=False) > 1).mean() * 100 for col in columnas_a_comparar
}).round(1)

print("Pozos físicos (siglas distintas):", df["sigla"].nunique(), "| registros (idpozo):", len(df))
pct_distinto.rename("% de grupos donde el valor cambia")

Pozos físicos (siglas distintas): 78352 | registros (idpozo): 85611


formacion                  97.8
geojson                    19.6
profundidad                 9.1
adjiv_fecha_inicio_perf    12.4
tipoestado                  9.2
empresa                     2.4
Name: % de grupos donde el valor cambia, dtype: float64

La **formación** cambia en casi todos los grupos, mientras que la ubicación, la profundidad y la fecha de perforación casi nunca cambian.
Un ejemplo: el mismo agujero, con la misma ubicación y la misma profundidad, atraviesa tres formaciones, y cada una tiene su propio `idpozo` y su propio estado.

In [21]:
df[df["sigla"] == "ACO.RN.TA-4002"][["idpozo", "sigla", "formacion", "profundidad", "adjiv_fecha_inicio_perf", "tipoestado"]]

,idpozo,sigla,formacion,profundidad,adjiv_fecha_inicio_perf,tipoestado
83326,164673,ACO.RN.TA-4002,quintuco,1458.0,2023-09-14,Parado Transitoriamente
83327,164674,ACO.RN.TA-4002,sierras blancas,1458.0,2023-09-14,Extracción Efectiva
83328,164675,ACO.RN.TA-4002,choiyoi,1458.0,2023-09-14,Parado Transitoriamente


**Conclusión:** `idpozo` identifica la combinación **pozo + formación**, no el pozo físico. El pozo físico se identifica por `sigla`.
H-08 no es, en su mayoría, un error: es cómo la fuente registra los pozos que producen de varias capas.

Queda un caso que esta explicación no cubre: filas que repiten **sigla y también formación**.
Ahí no hay una capa distinta que justifique el registro aparte, así que parecen **cargas duplicadas** → **H-23**.

In [22]:
misma_sigla_y_formacion = df.duplicated(["sigla", "formacion"], keep=False)
print("Filas con misma sigla y misma formación:", misma_sigla_y_formacion.sum(),
      "| siglas:", df.loc[misma_sigla_y_formacion, "sigla"].nunique())

df[misma_sigla_y_formacion].sort_values("sigla")[["idpozo", "sigla", "formacion", "empresa", "tipoestado"]].head(6)

Filas con misma sigla y misma formación: 353 | siglas: 159


,idpozo,sigla,formacion,empresa,tipoestado
4683,10229,AEC.RN.PP-9,punta rosada,EDHIPSA,En Estudio
4684,10230,AEC.RN.PP-9,punta rosada,EDHIPSA,En Estudio
61866,131411,APCO.Ch.CR.x-11,formación improductiva,INTEROIL ARGENTINA S A,A Abandonar
61876,131422,APCO.Ch.CR.x-11,formación improductiva,NaN,En Estudio
61860,131405,APCO.Ch.CR.x-7,formación improductiva,INTEROIL ARGENTINA S A,A Abandonar
61861,131406,APCO.Ch.CR.x-7,formación improductiva,NaN,No informado


### Cota: ¿hay valores imposiblemente altos?

En la sección 7 medimos las cotas en 0 o negativas (H-12), pero no las demasiado altas.
Comparamos la cota máxima de cada provincia con la **mediana** (el valor del medio) para ver si algún máximo se aleja demasiado de lo habitual.

In [23]:
df.groupby("provincia")["cota"].agg(["median", "max"]).round(0).sort_values("max", ascending=False).head(5)

,median,max
provincia,,
Neuquén,609.0,5543.0
Mendoza,942.0,2978.0
Rio Negro,364.0,2901.0
San Juan,975.0,1764.0
Salta,569.0,1299.0


En Neuquén aparece una cota de **5.543 m**, más alta que el volcán Domuyo (unos 4.700 m), el punto más alto de la provincia.
Es físicamente imposible → **H-24**.

La mediana de su yacimiento ronda los 590 m. El valor 5.543,41 se parece a 554,341 con la coma corrida:
es el mismo patrón que en las profundidades de H-11. Es una **hipótesis**, no se aplica sin confirmación de la fuente.

In [24]:
ALTURA_MAXIMA_NEUQUEN = 4700  # volcán Domuyo, el punto más alto de Neuquén (aproximado)

cota_imposible_alta = (df["provincia"] == "Neuquén") & (df["cota"] > ALTURA_MAXIMA_NEUQUEN)
mediana_por_yacimiento = df.groupby("yacimiento")["cota"].median()

df.loc[cota_imposible_alta, ["idpozo", "sigla", "yacimiento", "cota"]].assign(
    mediana_del_yacimiento=df["yacimiento"].map(mediana_por_yacimiento).round(0)
)

,idpozo,sigla,yacimiento,cota,mediana_del_yacimiento
72974,153315,YPF.Nq.Cto1-527,CAMPAMENTO UNO,5543.41,588.0


### Textos: ¿hay espacios sobrantes?

Para la computadora, `"GENERAL ROCA"` y `"GENERAL ROCA "` (con un espacio al final) son textos distintos.
Generamos una versión de las columnas de texto **sin espacios al principio ni al final y sin espacios dobles**
(solo para medir, en una variable aparte: `df` no se modifica) y contamos cuántos valores cambian.

In [25]:
columnas_texto = ["sigla", "area", "empresa", "yacimiento", "formacion"]
texto_original = df[columnas_texto]
texto_sin_espacios = texto_original.apply(lambda col: col.str.strip().str.replace(r"\s+", " ", regex=True))

cambia = texto_original.fillna("") != texto_sin_espacios.fillna("")
espacios_sobrantes = cambia.any(axis=1)
print("Filas con algún texto con espacios sobrantes:", espacios_sobrantes.sum())
cambia.sum().rename("valores con espacios sobrantes")

Filas con algún texto con espacios sobrantes: 5904


sigla          202
area          1074
empresa       4241
yacimiento     358
formacion       79
Name: valores con espacios sobrantes, dtype: int64

Hay espacios sobrantes en las cinco columnas → **H-25**.

¿Afecta a los hallazgos de códigos (H-17 y H-22)? Repetimos esas mediciones sobre el texto sin espacios sobrantes.

In [26]:
def nombres_con_varios_codigos(nombres, codigos):
    return (codigos.groupby(nombres).nunique() > 1).sum()

print("Áreas con más de un código:       texto original", nombres_con_varios_codigos(df["area"], df["cod_area"]),
      "| sin espacios sobrantes", nombres_con_varios_codigos(texto_sin_espacios["area"], df["cod_area"]))
print("Yacimientos con más de un código: texto original", nombres_con_varios_codigos(df["yacimiento"], df["cod_yacimiento"]),
      "| sin espacios sobrantes", nombres_con_varios_codigos(texto_sin_espacios["yacimiento"], df["cod_yacimiento"]))

Áreas con más de un código:       texto original 7 | sin espacios sobrantes 10
Yacimientos con más de un código: texto original 88 | sin espacios sobrantes 94


Los dos hallazgos **crecen**. Por ejemplo, `"BAJO DEL PICHE"` (con un código) y `"BAJO DEL PICHE "` (con otro)
se contaban como dos áreas distintas, cada una con un solo código. Sin el espacio son **una** área con **dos** códigos:
el espacio escondía el problema.

**Consecuencia para la limpieza:** quitar los espacios sobrantes tiene que ser la **primera** regla,
antes de las reglas sobre códigos (D-9). H-17 y H-22 quedan registrados con su valor sobre el texto original;
la cifra definitiva es la medida sin espacios sobrantes.

### Siglas: ¿se escribe siempre igual la misma sigla?

La sigla identifica al pozo físico (ver más arriba). Comparamos las siglas ya sin espacios sobrantes,
pasadas a mayúsculas: si dos escrituras distintas dan la misma **sigla normalizada**, la sigla se escribió de más de una forma.

In [27]:
sigla_normalizada = texto_sin_espacios["sigla"].str.upper()
escrituras_por_sigla = texto_sin_espacios["sigla"].groupby(sigla_normalizada).transform("nunique")
sigla_con_variantes = escrituras_por_sigla > 1

print("Filas con una sigla escrita de más de una forma:", sigla_con_variantes.sum(),
      "| siglas:", sigla_normalizada[sigla_con_variantes].nunique())
print("Pozos físicos: sigla original", df["sigla"].nunique(), "| sigla normalizada", sigla_normalizada.nunique())
print("Misma sigla y formación (H-23): sigla original", misma_sigla_y_formacion.sum(),
      "| sigla normalizada", pd.concat([sigla_normalizada, texto_sin_espacios["formacion"]], axis=1).duplicated(keep=False).sum())

df.loc[sigla_con_variantes, ["idpozo", "sigla", "formacion", "empresa"]].assign(
    sigla_normalizada=sigla_normalizada).sort_values("sigla_normalizada").head(6)

Filas con una sigla escrita de más de una forma: 85 | siglas: 37


Pozos físicos: sigla original 78352 | sigla normalizada 78299
Misma sigla y formación (H-23): sigla original 353 | sigla normalizada 369


,idpozo,sigla,formacion,empresa,sigla_normalizada
13842,51038,AEC.NQ.AC.a-2002,lotena,OILSTONE ENERGIA S.A.,AEC.NQ.AC.A-2002
76528,157635,AEC.Nq.AC.a-2002,precuyo,OILSTONE ENERGIA S.A.,AEC.NQ.AC.A-2002
76529,157636,AEC.Nq.AC.a-2002,basamento,OILSTONE ENERGIA S.A.,AEC.NQ.AC.A-2002
64114,133847,APA.NQ.BG-1064,centenario,BENTIA ENERGY S.A.,APA.NQ.BG-1064
76480,157587,APA.Nq.BG-1064,mulichinco,BENTIA ENERGY S.A.,APA.NQ.BG-1064
72731,153071,APA.NQ.BG-1080,mulichinco,BENTIA ENERGY S.A.,APA.NQ.BG-1080


La misma sigla aparece escrita, por ejemplo, con `NQ` y con `Nq` → **H-26**.
Con la sigla normalizada hay menos pozos físicos y más posibles cargas duplicadas (H-23).
La sigla original **no se modifica** (es un código oficial y sus minúsculas pueden tener significado);
para identificar al pozo físico se usa la sigla normalizada.

## 10. Registro de hallazgos

Cada problema detectado recibe un **código H-xx** y se mide en filas y porcentaje, según las dimensiones de la política de calidad.
Este registro es el punto de partida de `02_pozos_limpieza.ipynb`: cada regla de limpieza (R-xx) va a indicar qué hallazgo resuelve.

Se calcula con código para que, si el archivo raw se actualiza, el registro se actualice solo.
El **estado** de todos es *Pendiente*: en este notebook solo se diagnostica.

**H-19 a H-22** se agregaron el 28/09/2026, al construir el diccionario de datos (sección 9).
**H-23** se agregó el 30/09/2026, al investigar la decisión D-3.
**H-24** se agregó el 30/09/2026, al investigar la decisión D-6.
**H-25** y **H-26** se agregaron el 01/10/2026, al investigar la decisión D-8. H-17 y H-22 se miden sobre el texto original: sin espacios sobrantes, sus cifras suben (sección 9).

In [28]:
total = len(df)
codigos_repetidos = areas_con_varios_codigos[areas_con_varios_codigos > 1].index

hallazgos = [
    ("Completitud", "empresa vacía", df["empresa"].isna().sum()),
    ("Completitud", "formacion vacía", df["formacion"].isna().sum()),
    ("Completitud", "clasificacion y subclasificacion = 'No informado'", (df["clasificacion"] == "No informado").sum()),
    ("Completitud", "tipo_recurso = 'No informado'", (df["tipo_recurso"] == "No informado").sum()),
    ("Completitud", "sub_tipo_recurso = 'No informado'", (df["sub_tipo_recurso"] == "No informado").sum()),
    ("Completitud", "tipopozo, tipoextraccion y tipoestado = 'No informado'", (df["tipopozo"] == "No informado").sum()),
    ("Completitud", "fecha de inicio de perforación vacía", df["adjiv_fecha_inicio_perf"].isna().sum()),
    ("Unicidad", "sigla repetida con distinto idpozo", filas_sigla_repetida.sum()),
    ("Unicidad", "ubicación compartida con otro pozo", filas_ubicacion_repetida.sum()),
    ("Validez", "profundidad = 0", (df["profundidad"] == 0).sum()),
    ("Validez", "profundidad > 10.000 m", (df["profundidad"] > 10_000).sum()),
    ("Validez", "cota = 0 o negativa", (df["cota"] <= 0).sum()),
    ("Validez", "fecha 1900-01-01 (inicio de perforación)", (fechas["adjiv_fecha_inicio_perf"] == "1900-01-01").sum()),
    ("Validez", "fin de perforación antes del inicio",
     (fechas["adjiv_fecha_fin_perf"] < fechas["adjiv_fecha_inicio_perf"]).sum()),
    ("Validez", "fechas guardadas como texto (4 columnas)", total),
    ("Consistencia", "formacion en minúsculas (el resto de los textos en mayúsculas)", df["formacion"].str.islower().sum()),
    ("Consistencia", "pozos en áreas con más de un código de área", df["area"].isin(codigos_repetidos).sum()),
    ("Consistencia", "provincia = 'Estado Nacional' (no es una provincia)", (df["provincia"] == "Estado Nacional").sum()),
    ("Validez", "coordenadas con longitud y latitud invertidas", coordenadas_fuera.sum()),
    ("Completitud", "sub_tipo_recurso = 'No informado' en pozos CONVENCIONAL: significa 'no aplica' (corrige H-05)", subtipo_no_aplica.sum()),
    ("Consistencia", "pozo CONVENCIONAL con sub_tipo_recurso informado", convencional_con_subtipo.sum()),
    ("Consistencia", "pozos en yacimientos con más de un código", df["yacimiento"].isin(yacimientos_con_varios_codigos).sum()),
    ("Unicidad", "misma sigla y misma formación (posible carga duplicada)", misma_sigla_y_formacion.sum()),
    ("Validez", "cota mayor que la altura máxima de su provincia", cota_imposible_alta.sum()),
    ("Consistencia", "textos con espacios sobrantes (al principio, al final o dobles)", espacios_sobrantes.sum()),
    ("Consistencia", "sigla escrita de más de una forma (mayúsculas/minúsculas)", sigla_con_variantes.sum()),
]

registro_hallazgos = pd.DataFrame(hallazgos, columns=["dimension", "hallazgo", "filas"])
registro_hallazgos.insert(0, "id", [f"H-{i:02d}" for i in range(1, len(hallazgos) + 1)])
registro_hallazgos["pct"] = (registro_hallazgos["filas"] / total * 100).round(2)
registro_hallazgos["estado"] = "Pendiente"
registro_hallazgos

,id,dimension,hallazgo,filas,pct,estado
0,H-01,Completitud,empresa vacía,964,1.13,Pendiente
1,H-02,Completitud,formacion vacía,2815,3.29,Pendiente
2,H-03,Completitud,clasificacion y subclasificacion = 'No informado',18199,21.26,Pendiente
3,H-04,Completitud,tipo_recurso = 'No informado',21859,25.53,Pendiente
4,H-05,Completitud,sub_tipo_recurso = 'No informado',80608,94.16,Pendiente
5,H-06,Completitud,"tipopozo, tipoextraccion y tipoestado = 'No in...",55,0.06,Pendiente
6,H-07,Completitud,fecha de inicio de perforación vacía,34004,39.72,Pendiente
7,H-08,Unicidad,sigla repetida con distinto idpozo,13421,15.68,Pendiente
8,H-09,Unicidad,ubicación compartida con otro pozo,12104,14.14,Pendiente
9,H-10,Validez,profundidad = 0,10299,12.03,Pendiente


## 11. Conclusiones y decisiones pendientes

**Lo que está bien**
- No hay filas duplicadas ni `idpozo` repetidos: `idpozo` sirve como **clave única** para cruzar con los archivos de producción. Ojo: identifica un **pozo + formación**, no un pozo físico (sección 9).
- Las fechas que tienen valor se pueden convertir todas a fecha.

**Lo que hay que decidir en `02_pozos_limpieza.ipynb`** (cada decisión, con sus opciones, pros y contras, la toma la responsable del proyecto):

| Decisión | Hallazgos | Pregunta |
| --- | --- | --- |
| D-1 Nulos ocultos | H-03, H-04, H-05, H-06, H-20 | ¿Convertir "No informado" en nulo real, o dejarlo como categoría? ¿Cómo distinguir "no aplica" de "falta el dato"? |
| D-2 Datos faltantes | H-01, H-02, H-07 | ¿Marcar, completar o dejar vacíos? |
| D-3 Siglas repetidas | H-08, H-23 | Investigado en la sección 9: casi todas son un mismo pozo con varias formaciones. ¿Cómo distinguir pozo físico de registro? ¿Qué hacer con las posibles cargas duplicadas? |
| D-4 Ubicaciones compartidas | H-09 | ¿Plataformas con varios pozos o error de carga? |
| D-5 Profundidad | H-10, H-11 | ¿Marcar como sospechosos, corregir o dejar? El valor de 378.939 m es imposible. |
| D-6 Cota | H-12, H-24 | ¿Cota 0 o negativa es un dato real o un valor de relleno? ¿Qué hacer con la cota imposiblemente alta? |
| D-7 Fechas | H-13, H-14, H-15 | Convertir a tipo fecha; ¿qué hacer con 1900-01-01 y con los fines anteriores al inicio? |
| D-8 Textos | H-16, H-25, H-26 | ¿Unificar mayúsculas/minúsculas? ¿Quitar espacios sobrantes? ¿Cómo identificar al pozo físico si la sigla se escribe de varias formas? |
| D-9 Códigos | H-17, H-22 | ¿Cuál es la referencia: el nombre (del área o del yacimiento) o su código? |
| D-10 Jurisdicción | H-18 | "Estado Nacional" probablemente son pozos costa afuera (offshore). ¿Se deja o se recategoriza? |
| D-11 Coordenadas | H-19 | ¿Intercambiar longitud y latitud en los 2 pozos, o solo marcarlos? |
| D-12 Tipo y subtipo de recurso | H-21 | ¿El pozo CONVENCIONAL con subtipo TIGHT tiene mal el tipo o el subtipo? |

**Otra observación (no es un problema de calidad):** las columnas `geojson` y `geom` guardan la misma ubicación en dos formatos.
En la sección 9 se verificó que coinciden en todas las filas: evaluar en la limpieza si se conservan las dos.

> Criterio de la política: *ante la duda, marcar antes que eliminar*.
> `POLITICA_CALIDAD_DATOS.md` es la referencia para el diagnóstico y no se modifica;
> las reglas de limpieza (R-xx) se registran en `02_pozos_limpieza.ipynb`, cada una con el hallazgo (H-xx) que resuelve.